// @sanity/vision/lib/components/NoResultsDialog.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _ui = require("@sanity/ui");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function NoResultsDialog(props) {
  return /*#__PURE__*/_react.default.createElement(_ui.Box, {
    paddingY: 2,
    className: "vision_no-results"
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 5
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p",
    muted: true
  }, "No documents found in dataset ", /*#__PURE__*/_react.default.createElement("code", null, props.dataset), " that match query:"), /*#__PURE__*/_react.default.createElement(_ui.Code, null, props.query)));
}

NoResultsDialog.propTypes = {
  query: _propTypes.default.string.isRequired,
  dataset: _propTypes.default.string.isRequired
};
var _default = NoResultsDialog;
exports.default = _default;
