// @sanity/vision/lib/components/QueryErrorDialog.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _ui = require("@sanity/ui");

var _QueryErrorDetails = _interopRequireDefault(require("./QueryErrorDetails"));

var _QueryErrorDialog = require("./QueryErrorDialog.styled");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function QueryErrorDialog(props) {
  return /*#__PURE__*/_react.default.createElement(_ui.Box, {
    className: "vision_query-error"
  }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 5,
    marginTop: 2
  }, /*#__PURE__*/_react.default.createElement(_QueryErrorDialog.ErrorCode, {
    size: 1
  }, props.error.message), /*#__PURE__*/_react.default.createElement(_QueryErrorDetails.default, {
    error: props.error
  })));
}

QueryErrorDialog.propTypes = {
  error: _propTypes.default.instanceOf(Error)
};
var _default = QueryErrorDialog;
exports.default = _default;
