// @sanity/vision/lib/components/QueryEditor.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactCodemirror = require("react-codemirror2");

var _codemirror = _interopRequireDefault(require("codemirror"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _ui = require("@sanity/ui");

var _templateObject;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

require('codemirror/mode/javascript/javascript');

require('codemirror/addon/hint/show-hint');

require('codemirror/addon/edit/closebrackets');

var ReactCodeMirror = (0, _styledComponents.default)(_reactCodemirror.UnControlled)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  width: 100%;\n  box-sizing: border-box;\n  height: 100%;\n  overflow: hidden;\n  position: relative;\n\n  .CodeMirror {\n    width: 100%;\n    height: 100%;\n    overflow: hidden;\n    position: relative;\n  }\n\n  .CodeMirror-sizer {\n    padding-top: ", ";\n  }\n\n  .CodeMirror-linenumber {\n    padding: 0 3px 0 5px;\n    min-width: 1.5rem;\n    text-align: right;\n    color: var(--card-code-fg-color);\n    white-space: nowrap;\n  }\n"])), _ref => {
  var theme = _ref.theme;
  return (0, _ui.rem)(theme.sanity.space[5]);
});

class QueryEditor extends _react.default.PureComponent {
  constructor(props) {
    super(props);
    this.handleChange = this.handleChange.bind(this);
    this.getHint = this.getHint.bind(this);
  }

  getHint(cm, option) {
    var schema = this.props.schema;
    var cursor = cm.getCursor(); // const line = cm.getLine(cursor.line)

    var start = cursor.ch;
    var end = cursor.ch;
    var suggestions = [{
      text: '...',
      displayText: '... (Includes everything)'
    }];

    if (schema) {
      schema._original.types.forEach(type => {
        suggestions.push({
          text: type.name,
          displayText: "".concat(type.name, " (").concat(type.type, ") - ").concat(type.title)
        });
      });
    }

    return {
      list: suggestions,
      from: _codemirror.default.Pos(cursor.line, start),
      to: _codemirror.default.Pos(cursor.line, end)
    };
  }

  handleChange(editor, metadata, value) {
    this.props.onChange({
      query: value
    });
  }

  render() {
    var options = {
      theme: 'default CodeMirror-vision',
      lineNumbers: true,
      tabSize: 2,
      mode: {
        name: 'javascript',
        json: true
      },
      hintOptions: {
        hint: this.getHint
      },
      extraKeys: {
        Tab: false,
        'Ctrl-Space': 'autocomplete',
        'Ctrl-Enter': this.props.onExecute
      },
      autoCloseBrackets: true
    };
    return /*#__PURE__*/_react.default.createElement(ReactCodeMirror, {
      value: this.props.value,
      onChange: this.handleChange,
      options: options,
      className: this.props.className,
      onHeightChange: this.props.onHeightChange,
      autoCursor: false,
      autoScroll: true
    });
  }

}

QueryEditor.propTypes = {
  onExecute: _propTypes.default.func.isRequired,
  onChange: _propTypes.default.func.isRequired,
  value: _propTypes.default.string,
  schema: _propTypes.default.object,
  className: _propTypes.default.string,
  onHeightChange: _propTypes.default.func
};
QueryEditor.defaultProps = {
  className: 'vision_query-editor'
};
var _default = QueryEditor;
exports.default = _default;
