// @sanity/vision/lib/components/VisionGui.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _queryString = _interopRequireDefault(require("query-string"));

var _reactSplitPane = _interopRequireDefault(require("react-split-pane"));

var _icons = require("@sanity/icons");

var _isHotkey = _interopRequireDefault(require("is-hotkey"));

var _ui = require("@sanity/ui");

var _client = _interopRequireDefault(require("part:@sanity/base/client"));

var _components = require("@sanity/base/components");

var _vision = _interopRequireDefault(require("config:@sanity/vision"));

var _localState = require("../util/localState");

var _parseApiQueryString = _interopRequireDefault(require("../util/parseApiQueryString"));

var _prefixApiVersion = _interopRequireDefault(require("../util/prefixApiVersion"));

var _tryParseParams = _interopRequireDefault(require("../util/tryParseParams"));

var _encodeQueryString = _interopRequireDefault(require("../util/encodeQueryString"));

var _apiVersions = require("../apiVersions");

var _resizeObserver = require("../util/resizeObserver");

var _DelayedSpinner = _interopRequireDefault(require("./DelayedSpinner"));

var _QueryEditor = _interopRequireDefault(require("./QueryEditor"));

var _ParamsEditor = _interopRequireDefault(require("./ParamsEditor"));

var _ResultView = _interopRequireDefault(require("./ResultView"));

var _NoResultsDialog = _interopRequireDefault(require("./NoResultsDialog"));

var _QueryErrorDialog = _interopRequireDefault(require("./QueryErrorDialog"));

var _VisionGui = require("./VisionGui.styled");

require("codemirror/lib/codemirror.css?raw");

require("codemirror/theme/material.css?raw");

require("codemirror/addon/hint/show-hint.css?raw");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function nodeContains(node, other) {
  if (!node || !other) {
    return false;
  } // eslint-disable-next-line no-bitwise


  return node === other || !!(node.compareDocumentPosition(other) & 16);
}
/* eslint-enable import/no-unassigned-import, import/no-unresolved */


var sanityUrl = /\.api\.sanity\.io\/(vx|v1|v\d{4}-\d\d-\d\d)\/.*?(?:query|listen)\/(.*?)\?(.*)/;

var isRunHotkey = event => (0, _isHotkey.default)('ctrl+enter', event) || (0, _isHotkey.default)('mod+enter', event);

var handleCopyUrl = () => {
  var emailLink = document.querySelector('#vision-query-url');
  emailLink.select();

  try {
    document.execCommand('copy');
  } catch (err) {
    // eslint-disable-next-line no-console
    console.error('Unable to copy to clipboard :(');
  }
};

function calculatePaneSizeOptions(rootHeight) {
  return {
    defaultSize: rootHeight / 2,
    size: rootHeight > 550 ? undefined : rootHeight * 0.4,
    allowResize: rootHeight > 550,
    minSize: Math.min(170, Math.max(170, rootHeight / 2)),
    maxSize: rootHeight > 650 ? rootHeight * 0.7 : rootHeight * 0.6
  };
}

var DEFAULT_API_VERSION = '2021-03-25';

class VisionGui extends _react.default.PureComponent {
  constructor(props) {
    super(props);
    var lastQuery = (0, _localState.getState)('lastQuery');
    var lastParams = (0, _localState.getState)('lastParams');
    var firstDataset = this.props.datasets[0] && this.props.datasets[0].name;
    var defaultDataset = _client.default.config().dataset || firstDataset;
    var defaultApiVersion = (0, _prefixApiVersion.default)("".concat(_vision.default.defaultApiVersion || DEFAULT_API_VERSION));
    var dataset = (0, _localState.getState)('dataset', defaultDataset);
    var apiVersion = (0, _localState.getState)('apiVersion', defaultApiVersion);

    if (!this.props.datasets.some(_ref => {
      var name = _ref.name;
      return name === dataset;
    })) {
      dataset = defaultDataset;
    }

    if (!_apiVersions.apiVersions.includes(apiVersion)) {
      apiVersion = _apiVersions.apiVersions[0];
    }

    this._visionRoot = /*#__PURE__*/_react.default.createRef();
    this._resizeListener = /*#__PURE__*/_react.default.createRef();
    this._queryEditorContainer = /*#__PURE__*/_react.default.createRef();
    this._paramsEditorContainer = /*#__PURE__*/_react.default.createRef();
    this.client = _client.default.withConfig({
      apiVersion,
      dataset
    });
    this.subscribers = {}; // Initial root height without header

    var bodyHeight = document.body.getBoundingClientRect().height - 60;
    this.state = {
      paramValidationMarkers: [],
      validParams: true,
      query: lastQuery,
      params: lastParams && (0, _tryParseParams.default)(lastParams),
      rawParams: lastParams,
      queryInProgress: false,
      paneSizeOptions: calculatePaneSizeOptions(bodyHeight),
      dataset,
      apiVersion
    };
    this.handleChangeDataset = this.handleChangeDataset.bind(this);
    this.handleChangeApiVersion = this.handleChangeApiVersion.bind(this);
    this.handleCustomApiVersionChange = this.handleCustomApiVersionChange.bind(this);
    this.handleListenExecution = this.handleListenExecution.bind(this);
    this.handleListenerMutation = this.handleListenerMutation.bind(this);
    this.handleQueryExecution = this.handleQueryExecution.bind(this);
    this.handleQueryChange = this.handleQueryChange.bind(this);
    this.handleParamsChange = this.handleParamsChange.bind(this);
    this.handlePaste = this.handlePaste.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);
    this.handleResize = this.handleResize.bind(this);
  }

  componentDidMount() {
    window.document.addEventListener('paste', this.handlePaste);
    window.document.addEventListener('keydown', this.handleKeyDown);
    this.handleResizeListen();
  }

  componentWillUnmount() {
    this.cancelQuery();
    this.cancelListener();
    this.cancelEventListener();
    this.cancelResizeListener();
  }

  handleResizeListen() {
    this._resizeListener.current = new _resizeObserver.ResizeObserver(this.handleResize);

    this._resizeListener.current.observe(this._visionRoot.current);
  }

  handleResize(entries) {
    var entry = entries === null || entries === void 0 ? void 0 : entries[0];
    this.setState(prevState => _objectSpread(_objectSpread({}, prevState), {}, {
      paneSizeOptions: calculatePaneSizeOptions(entry.contentRect.height)
    }));
  }

  cancelResizeListener() {
    if (this._resizeListener.current) {
      this._resizeListener.current.disconnect();
    }
  }

  handlePaste(evt) {
    var data = evt.clipboardData.getData('text/plain');
    var match = data.match(sanityUrl);

    if (!match) {
      return;
    }

    var _match = _slicedToArray(match, 4),
        apiVersion = _match[1],
        dataset = _match[2],
        urlQuery = _match[3];

    var qs = _queryString.default.parse(urlQuery);

    var parts;

    try {
      parts = (0, _parseApiQueryString.default)(qs);
    } catch (err) {
      console.warn('Error while trying to parse API URL: ', err.message); // eslint-disable-line no-console

      return; // Give up on error
    }

    if (this.state.data !== dataset) {
      (0, _localState.storeState)('dataset', dataset);
    }

    if (this.state.apiVersion !== apiVersion) {
      (0, _localState.storeState)('apiVersion', apiVersion);
    }

    evt.preventDefault();
    this.client.config({
      dataset,
      apiVersion
    });
    this.setState({
      dataset,
      apiVersion,
      query: parts.query,
      params: parts.params,
      rawParams: JSON.stringify(parts.params, null, 2)
    }, () => {
      this.handleQueryExecution();
    });
  }

  cancelQuery() {
    if (!this.subscribers.query) {
      return;
    }

    this.subscribers.query.unsubscribe();
    this.subscribers.query = null;
  }

  cancelListener() {
    if (!this.subscribers.listen) {
      return;
    }

    this.subscribers.listen.unsubscribe();
    this.subscribers.listen = null;
  }

  cancelEventListener() {
    window.removeEventListener('keydown', this.handleKeyDown);
  }

  handleChangeDataset(evt) {
    var dataset = evt.target.value;
    (0, _localState.storeState)('dataset', dataset);
    this.setState({
      dataset
    });
    this.client.config({
      dataset
    });
    this.handleQueryExecution();
  }

  handleChangeApiVersion(evt) {
    var apiVersion = evt.target.value;

    if (apiVersion === 'other') {
      this.setState({
        customApiVersion: true
      });
      return;
    }

    (0, _localState.storeState)('apiVersion', apiVersion);
    this.setState({
      apiVersion,
      customApiVersion: undefined
    });
    this.client.config({
      apiVersion
    });
    this.handleQueryExecution();
  }

  handleCustomApiVersionChange(evt) {
    var customApiVersion = evt.target.value;
    var parseableApiVersion = customApiVersion.replace(/^v/, '').trim().match(/^\d{4}-\d{2}-\d{2}$/);
    var isValidApiVersion = !isNaN(Date.parse(parseableApiVersion));
    this.setState(prevState => ({
      apiVersion: isValidApiVersion ? customApiVersion : prevState.apiVersion,
      customApiVersion: customApiVersion || true,
      isValidApiVersion
    }), () => {
      if (!this.state.isValidApiVersion) {
        return;
      }

      this.client.config({
        apiVersion: this.state.customApiVersion
      });
    });
  }

  handleListenerMutation(mut) {
    // eslint-disable-next-line react/no-access-state-in-setstate
    var listenMutations = [mut].concat(this.state.listenMutations);

    if (listenMutations.length > 50) {
      listenMutations.pop();
    }

    this.setState({
      listenMutations
    });
  }

  handleKeyDown(event) {
    var validParams = this.state.validParams;
    var isWithinRoot = this._visionRoot.current && nodeContains(this._visionRoot.current, event.target);

    if (isRunHotkey(event) && isWithinRoot && validParams) {
      this.handleQueryExecution();
    }
  }

  handleListenExecution() {
    var _this$state = this.state,
        query = _this$state.query,
        params = _this$state.params,
        rawParams = _this$state.rawParams,
        listenInProgress = _this$state.listenInProgress;

    if (listenInProgress) {
      this.cancelListener();
      this.setState({
        listenInProgress: false
      });
      return;
    }

    var paramsError = params instanceof Error && params;
    var url = this.client.getUrl(this.client.getDataUrl('listen', (0, _encodeQueryString.default)(query, params)));
    (0, _localState.storeState)('lastQuery', query);
    (0, _localState.storeState)('lastParams', rawParams);
    this.cancelQuery();
    this.setState({
      url,
      listenMutations: [],
      queryInProgress: false,
      result: undefined,
      listenInProgress: !paramsError && Boolean(query),
      error: paramsError || undefined,
      // result: undefined,
      queryTime: null,
      e2eTime: null
    });

    if (!query || paramsError) {
      return;
    }

    this.subscribers.listen = this.client.listen(query, params, {}).subscribe({
      next: this.handleListenerMutation,
      error: _error => this.setState({
        error: _error,
        query,
        listenInProgress: false
      })
    });
  }

  handleQueryExecution() {
    var _this$state2 = this.state,
        query = _this$state2.query,
        params = _this$state2.params,
        rawParams = _this$state2.rawParams;
    var paramsError = params instanceof Error && params;
    (0, _localState.storeState)('lastQuery', query);
    (0, _localState.storeState)('lastParams', rawParams);
    this.cancelListener();
    this.setState({
      queryInProgress: !paramsError && Boolean(query),
      listenInProgress: false,
      listenMutations: [],
      error: paramsError || undefined,
      result: undefined,
      queryTime: null,
      e2eTime: null
    });

    if (!query || paramsError) {
      return;
    }

    var url = this.client.getUrl(this.client.getDataUrl('query', (0, _encodeQueryString.default)(query, params)));
    this.setState({
      url
    });
    var queryStart = Date.now();
    this.subscribers.query = this.client.observable.fetch(query, params, {
      filterResponse: false,
      tag: 'vision'
    }).subscribe({
      next: res => {
        this.setState({
          executedQuery: query,
          queryTime: res.ms,
          e2eTime: Date.now() - queryStart,
          result: res.result,
          queryInProgress: false,
          error: null
        });
      },
      error: _error2 => this.setState({
        error: _error2,
        query,
        queryInProgress: false
      })
    });
  }

  handleQueryChange(data) {
    this.setState({
      query: data.query
    });
  }

  handleParamsChange(data) {
    var raw = data.raw,
        parsed = data.parsed,
        valid = data.valid,
        validationError = data.validationError;
    var paramValidationMarkers = valid ? [] : [{
      type: 'validation',
      level: 'error',
      item: {
        message: validationError || 'Invalid JSON'
      }
    }];
    this.setState({
      rawParams: raw,
      params: parsed,
      validParams: valid,
      paramValidationMarkers
    });
  }

  render() {
    var datasets = this.props.datasets;
    var _this$state3 = this.state,
        error = _this$state3.error,
        result = _this$state3.result,
        url = _this$state3.url,
        query = _this$state3.query,
        queryInProgress = _this$state3.queryInProgress,
        executedQuery = _this$state3.executedQuery,
        listenInProgress = _this$state3.listenInProgress,
        paneSizeOptions = _this$state3.paneSizeOptions,
        queryTime = _this$state3.queryTime,
        e2eTime = _this$state3.e2eTime,
        listenMutations = _this$state3.listenMutations,
        apiVersion = _this$state3.apiVersion,
        dataset = _this$state3.dataset,
        customApiVersion = _this$state3.customApiVersion,
        isValidApiVersion = _this$state3.isValidApiVersion,
        validParams = _this$state3.validParams,
        paramValidationMarkers = _this$state3.paramValidationMarkers;
    var hasResult = !error && !queryInProgress && typeof result !== 'undefined';
    var hasEmptyResult = hasResult && Array.isArray(result) && result.length === 0; // Note that because of react-json-inspector, we need at least one
    // addressable, non-generated class name. Therefore;
    // leave `sanity-vision` untouched!

    var visionClass = ['sanity-vision'].filter(Boolean).join(' ');
    return /*#__PURE__*/_react.default.createElement(_VisionGui.Root, {
      tone: "default",
      className: visionClass,
      ref: this._visionRoot,
      display: "flex",
      sizing: "border",
      overflow: "hidden"
    }, /*#__PURE__*/_react.default.createElement(_VisionGui.GlobalCodeMirrorStyle, null), /*#__PURE__*/_react.default.createElement(_VisionGui.Header, {
      paddingX: 3,
      paddingY: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Grid, {
      columns: [6, 6, 12]
    }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 1,
      column: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Stack, null, /*#__PURE__*/_react.default.createElement(_ui.Card, {
      paddingY: 2
    }, /*#__PURE__*/_react.default.createElement(_VisionGui.StyledLabel, null, "Dataset")), /*#__PURE__*/_react.default.createElement(_ui.Select, {
      value: dataset,
      onChange: this.handleChangeDataset
    }, datasets.map(ds => /*#__PURE__*/_react.default.createElement("option", {
      key: ds.name
    }, ds.name))))), /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 1,
      column: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Stack, null, /*#__PURE__*/_react.default.createElement(_ui.Card, {
      paddingY: 2
    }, /*#__PURE__*/_react.default.createElement(_VisionGui.StyledLabel, null, "API version")), /*#__PURE__*/_react.default.createElement(_ui.Select, {
      value: customApiVersion ? 'other' : apiVersion,
      onChange: this.handleChangeApiVersion
    }, _apiVersions.apiVersions.map(version => /*#__PURE__*/_react.default.createElement("option", {
      key: version
    }, version)), /*#__PURE__*/_react.default.createElement("option", {
      key: "other",
      value: "other"
    }, "Other")))), customApiVersion && /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 1,
      column: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Stack, null, /*#__PURE__*/_react.default.createElement(_ui.Card, {
      paddingY: 2
    }, /*#__PURE__*/_react.default.createElement(_VisionGui.StyledLabel, null, "Custom API version")), /*#__PURE__*/_react.default.createElement(_ui.TextInput, {
      value: typeof customApiVersion === 'string' ? customApiVersion : '',
      onChange: this.handleCustomApiVersionChange,
      customValidity: isValidApiVersion ? undefined : 'Invalid API version'
    }))), typeof url === 'string' ? /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 1,
      flex: 1,
      column: customApiVersion ? 6 : 8
    }, /*#__PURE__*/_react.default.createElement(_ui.Stack, null, /*#__PURE__*/_react.default.createElement(_ui.Card, {
      paddingY: 2
    }, /*#__PURE__*/_react.default.createElement(_VisionGui.StyledLabel, null, "Query URL\xA0", /*#__PURE__*/_react.default.createElement(_VisionGui.QueryCopyLink, {
      onClick: handleCopyUrl
    }, "[copy]"))), /*#__PURE__*/_react.default.createElement(_ui.TextInput, {
      readOnly: true,
      id: "vision-query-url",
      value: url,
      iconRight: _icons.CopyIcon,
      onClick: handleCopyUrl
    }))) : /*#__PURE__*/_react.default.createElement(_ui.Box, {
      flex: 1
    }))), /*#__PURE__*/_react.default.createElement(_VisionGui.SplitpaneContainer, {
      flex: 1
    }, /*#__PURE__*/_react.default.createElement(_reactSplitPane.default, {
      split: "vertical",
      minSize: 280,
      defaultSize: 400,
      maxSize: -400
    }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
      height: "stretch",
      flex: 1
    }, /*#__PURE__*/_react.default.createElement(_reactSplitPane.default, {
      className: "sidebarPanes",
      split: "horizontal",
      defaultSize: paneSizeOptions.defaultSize,
      size: paneSizeOptions.size,
      allowResize: paneSizeOptions.allowResize,
      minSize: paneSizeOptions.minSize,
      maxSize: paneSizeOptions.maxSize,
      primary: "first"
    }, /*#__PURE__*/_react.default.createElement(_VisionGui.InputContainer, {
      display: "flex",
      ref: this._queryEditorContainer
    }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
      flex: 1
    }, /*#__PURE__*/_react.default.createElement(_VisionGui.InputBackgroundContainerLeft, null, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
      marginLeft: 3
    }, /*#__PURE__*/_react.default.createElement(_VisionGui.StyledLabel, {
      muted: true
    }, "Query"))), /*#__PURE__*/_react.default.createElement(_QueryEditor.default, {
      value: this.state.query,
      onExecute: this.handleQueryExecution,
      onChange: this.handleQueryChange,
      schema: this.props.schema
    }))), /*#__PURE__*/_react.default.createElement(_VisionGui.InputContainer, {
      display: "flex",
      ref: this._paramsEditorContainer
    }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
      flex: 1,
      tone: validParams ? 'default' : 'critical'
    }, /*#__PURE__*/_react.default.createElement(_VisionGui.InputBackgroundContainerLeft, null, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
      marginLeft: 3
    }, /*#__PURE__*/_react.default.createElement(_VisionGui.StyledLabel, {
      muted: true
    }, "Params"), paramValidationMarkers.length > 0 && /*#__PURE__*/_react.default.createElement(_ui.Box, {
      marginLeft: 2
    }, /*#__PURE__*/_react.default.createElement(_components.FormFieldValidationStatus, {
      fontSize: 1,
      __unstable_markers: paramValidationMarkers
    })))), /*#__PURE__*/_react.default.createElement(_ParamsEditor.default, {
      value: this.state.rawParams,
      onExecute: this.handleQueryExecution,
      onChange: this.handleParamsChange
    })), /*#__PURE__*/_react.default.createElement(_VisionGui.ControlsContainer, null, /*#__PURE__*/_react.default.createElement(_ui.Card, {
      padding: 3,
      paddingX: 3
    }, /*#__PURE__*/_react.default.createElement(_ui.Tooltip, {
      content: /*#__PURE__*/_react.default.createElement(_ui.Card, {
        padding: 2,
        radius: 4
      }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
        size: 1,
        muted: true
      }, "Parameters are not valid JSON")),
      placement: "top",
      disabled: validParams,
      portal: true
    }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
      justify: "space-evenly"
    }, /*#__PURE__*/_react.default.createElement(_ui.Card, {
      flex: 1
    }, /*#__PURE__*/_react.default.createElement(_ui.Tooltip, {
      content: /*#__PURE__*/_react.default.createElement(_ui.Card, {
        padding: 2,
        radius: 4
      }, /*#__PURE__*/_react.default.createElement(_ui.Hotkeys, {
        keys: ['Ctrl', 'Enter']
      })),
      placement: "top",
      portal: true
    }, /*#__PURE__*/_react.default.createElement(_VisionGui.ButtonFullWidth, {
      onClick: this.handleQueryExecution,
      icon: _icons.PlayIcon,
      loading: queryInProgress,
      disabled: listenInProgress || !validParams,
      tone: "primary",
      text: "Fetch"
    }))), /*#__PURE__*/_react.default.createElement(_ui.Card, {
      flex: 1,
      marginLeft: 3
    }, /*#__PURE__*/_react.default.createElement(_VisionGui.ButtonFullWidth, {
      onClick: this.handleListenExecution,
      type: "button",
      icon: listenInProgress ? _icons.StopIcon : _icons.PlayIcon,
      text: listenInProgress ? 'Stop' : 'Listen',
      mode: "ghost",
      disabled: !validParams,
      tone: listenInProgress ? 'positive' : 'default'
    }))))))))), /*#__PURE__*/_react.default.createElement(_VisionGui.ResultOuterContainer, {
      direction: "column"
    }, /*#__PURE__*/_react.default.createElement(_VisionGui.ResultInnerContainer, {
      flex: 1
    }, /*#__PURE__*/_react.default.createElement(_VisionGui.ResultContainer, {
      flex: 1,
      overflow: "hidden",
      tone: error ? 'critical' : 'default',
      $isInvalid: error
    }, /*#__PURE__*/_react.default.createElement(_VisionGui.Result, {
      padding: 3,
      paddingTop: 5,
      overflow: "auto",
      $isInvalid: error
    }, /*#__PURE__*/_react.default.createElement(_VisionGui.InputBackgroundContainer, null, /*#__PURE__*/_react.default.createElement(_ui.Box, {
      marginLeft: 3
    }, /*#__PURE__*/_react.default.createElement(_VisionGui.StyledLabel, {
      muted: true
    }, "Result"))), queryInProgress && /*#__PURE__*/_react.default.createElement(_ui.Box, {
      marginTop: 3
    }, /*#__PURE__*/_react.default.createElement(_DelayedSpinner.default, null)), error && /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_QueryErrorDialog.default, {
      error: error
    })), hasResult && !hasEmptyResult && /*#__PURE__*/_react.default.createElement(_ResultView.default, {
      data: result,
      query: query
    }), hasEmptyResult && /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_NoResultsDialog.default, {
      query: executedQuery,
      dataset: dataset
    })), listenMutations && listenMutations.length > 0 && /*#__PURE__*/_react.default.createElement(_ResultView.default, {
      data: listenMutations
    })))), /*#__PURE__*/_react.default.createElement(_VisionGui.TimingsFooter, null, /*#__PURE__*/_react.default.createElement(_VisionGui.TimingsCard, {
      paddingX: 4,
      paddingY: 3,
      sizing: "border"
    }, /*#__PURE__*/_react.default.createElement(_VisionGui.TimingsTextContainer, {
      align: "center"
    }, /*#__PURE__*/_react.default.createElement(_ui.Card, null, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      muted: true
    }, "Execution: ", typeof queryTime === 'number' ? "".concat(queryTime, "ms") : 'n/a')), /*#__PURE__*/_react.default.createElement(_ui.Card, {
      marginLeft: 4
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      muted: true
    }, "End-to-end: ", typeof e2eTime === 'number' ? "".concat(e2eTime, "ms") : 'n/a')))))))));
  }

}

VisionGui.propTypes = {
  schema: _propTypes.default.object,
  datasets: _propTypes.default.arrayOf(_propTypes.default.shape({
    name: _propTypes.default.string
  }))
};
var _default = VisionGui;
exports.default = _default;
