// @sanity/form-builder/lib/inputs/PortableText/hooks/useObjectEditData.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useObjectEditData = useObjectEditData;

var _portableTextEditor = require("@sanity/portable-text-editor");

var _types = require("@sanity/types");

var _react = require("react");

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

// This hook will gather the info we need to create a edit modal for some embedded object in the editor
// based on the current focusPath details.
function useObjectEditData(focusPath, refs) {
  var editor = (0, _portableTextEditor.usePortableTextEditor)();
  var selection = (0, _portableTextEditor.usePortableTextEditorSelection)();
  return (0, _react.useMemo)(() => {
    var blockSegment = focusPath && (0, _types.isKeySegment)(focusPath[0]) ? focusPath[0] : undefined;
    var isChild = blockSegment && focusPath && focusPath[1] === 'children' && (0, _types.isKeyedObject)(focusPath[2]);
    var isAnnotation = blockSegment && focusPath[1] === 'markDefs'; // Handle focusPath pointing to a annotated span

    if (isAnnotation) {
      var _PortableTextEditor$f = _portableTextEditor.PortableTextEditor.findByPath(editor, focusPath.slice(0, 1)),
          _PortableTextEditor$f2 = _slicedToArray(_PortableTextEditor$f, 1),
          node = _PortableTextEditor$f2[0];

      var block = node ? node : undefined;
      var markDefSegment = block && _portableTextEditor.PortableTextEditor.isVoid(editor, block) === false && focusPath[2];

      if (markDefSegment) {
        var span = block.children.find(child => Array.isArray(child.marks) && child.marks.includes(markDefSegment._key));

        if (span) {
          var spanPath = [blockSegment, 'children', {
            _key: span._key
          }];
          return {
            editorPath: spanPath,
            formBuilderPath: focusPath.slice(0, 3),
            returnToSelection: selection,
            kind: 'annotation',
            editorHTMLElementRef: refs.child
          };
        }
      }
    } // Handle focusPath pointing to block objects or inline objects


    if (focusPath && (isChild && focusPath.length > 3 || !isChild && focusPath.length > 1)) {
      var kind = 'blockObject';
      var path = focusPath.slice(0, 1);

      if (isChild) {
        kind = 'inlineObject';
        path = path.concat(focusPath.slice(1, 3));
      }

      var _PortableTextEditor$f3 = _portableTextEditor.PortableTextEditor.findByPath(editor, path),
          _PortableTextEditor$f4 = _slicedToArray(_PortableTextEditor$f3, 1),
          _node = _PortableTextEditor$f4[0]; // Only if it actually exists


      if (_node) {
        return {
          editorPath: path,
          formBuilderPath: path,
          kind,
          returnToSelection: selection,
          editorHTMLElementRef: isChild ? refs.inline : refs.block
        };
      }
    }

    refs.child.current = null;
    refs.block.current = null;
    return null;
  }, [editor, focusPath, refs.block, refs.child, refs.inline, selection]);
}
