// @sanity/desk-tool/lib/components/confirmDeleteDialog/ConfirmDeleteDialogBody.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ConfirmDeleteDialogBody = ConfirmDeleteDialogBody;

var _react = _interopRequireDefault(require("react"));

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _preview = require("@sanity/base/preview");

var _reactCopyToClipboard = require("react-copy-to-clipboard");

var _ConfirmDeleteDialogBody = require("./ConfirmDeleteDialogBody.styles");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function getSchemaType(typeName) {
  var schemaMod = require('part:@sanity/base/schema');

  var schema = schemaMod.default || schemaMod;
  return schema.get(typeName) || null;
}
/**
 * The inner part of the `ConfirmDeleteDialog`. This is ran when both the
 * `crossDatasetReferences` and `internalReferences` are loaded.
 */


function ConfirmDeleteDialogBody(_ref) {
  var crossDatasetReferences = _ref.crossDatasetReferences,
      internalReferences = _ref.internalReferences,
      documentTitle = _ref.documentTitle,
      totalCount = _ref.totalCount,
      action = _ref.action,
      projectIds = _ref.projectIds;
  var toast = (0, _ui.useToast)();

  if ((internalReferences === null || internalReferences === void 0 ? void 0 : internalReferences.totalCount) === 0 && (crossDatasetReferences === null || crossDatasetReferences === void 0 ? void 0 : crossDatasetReferences.totalCount) === 0) {
    return /*#__PURE__*/_react.default.createElement(_ui.Text, {
      as: "p"
    }, "Are you sure you want to delete ", /*#__PURE__*/_react.default.createElement("strong", null, "\u201C", documentTitle, "\u201D"), "?");
  }

  var documentCount = crossDatasetReferences.totalCount === 1 ? '1 document' : "".concat(crossDatasetReferences.totalCount.toLocaleString(), " documents");
  var projectCount = projectIds.length === 1 ? 'another project' : "".concat(projectIds.length, " projects");
  var projectIdList = "Project ID".concat(projectIds.length === 1 ? '' : 's', ": ").concat(projectIds.join(', '));
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_ui.Card, {
    padding: 3,
    radius: 2,
    tone: "caution",
    marginBottom: 4,
    flex: "none"
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, null, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    "aria-hidden": "true",
    size: 1
  }, /*#__PURE__*/_react.default.createElement(_icons.WarningOutlineIcon, null)), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1,
    marginLeft: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1
  }, totalCount === 1 ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "1 document refers to \u201C", documentTitle, "\u201D") : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, totalCount.toLocaleString(), " documents refer to \u201C", documentTitle, "\u201D"))))), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: "none",
    marginBottom: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, null, "You may not be able to ", action, " \u201C", documentTitle, "\u201D because the following documents refer to it:")), /*#__PURE__*/_react.default.createElement(_ConfirmDeleteDialogBody.ReferencesCard, null, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    direction: "column",
    height: "fill"
  }, internalReferences.totalCount > 0 && /*#__PURE__*/_react.default.createElement(_ConfirmDeleteDialogBody.InternalReferences, {
    "data-testid": "internal-references"
  }, internalReferences.references.map(internalReference => {
    var type = getSchemaType(internalReference._type);
    return /*#__PURE__*/_react.default.createElement(_ui.Box, {
      as: "li",
      key: internalReference._id,
      paddingX: 3,
      paddingY: 3
    }, type ? /*#__PURE__*/_react.default.createElement(_preview.SanityPreview, {
      type: type,
      value: internalReference,
      layout: "default"
    }) : /*#__PURE__*/_react.default.createElement(_preview.SanityDefaultPreview, {
      value: {
        title: 'Preview Unavailable',
        subtitle: "ID: ".concat(internalReference._id),
        media: /*#__PURE__*/_react.default.createElement(_icons.UnknownIcon, null)
      },
      layout: "default"
    }));
  }), internalReferences.totalCount > internalReferences.references.length && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    as: "li",
    padding: 3
  }, /*#__PURE__*/_react.default.createElement(_ConfirmDeleteDialogBody.OtherReferenceCount, internalReferences))), crossDatasetReferences.totalCount > 0 && /*#__PURE__*/_react.default.createElement(_ConfirmDeleteDialogBody.CrossDatasetReferencesDetails, {
    "data-testid": "cross-dataset-references",
    style: {
      // only add the border if needed
      borderTop: internalReferences.totalCount > 0 ? '1px solid var(--card-shadow-outline-color)' : undefined
    }
  }, /*#__PURE__*/_react.default.createElement(_ConfirmDeleteDialogBody.CrossDatasetReferencesSummary, null, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    padding: 4,
    align: "center"
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginRight: 4
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 3
  }, /*#__PURE__*/_react.default.createElement(_icons.DocumentsIcon, null))), /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    marginRight: 4,
    direction: "column"
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginBottom: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, null, documentCount, " in ", projectCount)), /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    title: projectIdList,
    textOverflow: "ellipsis",
    size: 1,
    muted: true
  }, projectIdList))), /*#__PURE__*/_react.default.createElement(_ConfirmDeleteDialogBody.ChevronWrapper, null, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    muted: true
  }, /*#__PURE__*/_react.default.createElement(_icons.ChevronDownIcon, null))))), /*#__PURE__*/_react.default.createElement(_ConfirmDeleteDialogBody.TableContainer, null, /*#__PURE__*/_react.default.createElement(_ConfirmDeleteDialogBody.Table, null, /*#__PURE__*/_react.default.createElement("thead", null, /*#__PURE__*/_react.default.createElement("tr", null, /*#__PURE__*/_react.default.createElement("th", null, /*#__PURE__*/_react.default.createElement(_ui.Label, {
    muted: true,
    size: 0
  }, "Project ID")), /*#__PURE__*/_react.default.createElement("th", null, /*#__PURE__*/_react.default.createElement(_ui.Label, {
    muted: true,
    size: 0
  }, "Dataset")), /*#__PURE__*/_react.default.createElement("th", null, /*#__PURE__*/_react.default.createElement(_ui.Label, {
    muted: true,
    size: 0
  }, "Document ID")))), /*#__PURE__*/_react.default.createElement("tbody", null, crossDatasetReferences.references.filter(reference => {
    return 'projectId' in reference && 'datasetName' in reference && 'documentId' in reference;
  }).map((_ref2, index) => {
    var projectId = _ref2.projectId,
        datasetName = _ref2.datasetName,
        documentId = _ref2.documentId;
    return (
      /*#__PURE__*/
      // eslint-disable-next-line react/no-array-index-key
      _react.default.createElement("tr", {
        key: "".concat(documentId, "-").concat(index)
      }, /*#__PURE__*/_react.default.createElement("td", null, /*#__PURE__*/_react.default.createElement(_ui.Text, {
        size: 1
      }, projectId)), /*#__PURE__*/_react.default.createElement("td", null, /*#__PURE__*/_react.default.createElement(_ui.Text, {
        size: 1
      }, datasetName)), /*#__PURE__*/_react.default.createElement("td", null, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
        align: "center",
        gap: 2,
        justify: "flex-end"
      }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
        textOverflow: "ellipsis",
        size: 1
      }, documentId), /*#__PURE__*/_react.default.createElement(_reactCopyToClipboard.CopyToClipboard, {
        text: documentId // eslint-disable-next-line react/jsx-no-bind
        ,
        onCopy: () => {
          // TODO: this isn't visible with the dialog open
          toast.push({
            title: 'Copied document ID to clipboard!',
            status: 'success'
          });
        }
      }, /*#__PURE__*/_react.default.createElement(_ui.Button, {
        title: "Copy ID to clipboard",
        mode: "bleed",
        icon: _icons.ClipboardIcon,
        fontSize: 0
      })))))
    );
  }))), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 2
  }, /*#__PURE__*/_react.default.createElement(_ConfirmDeleteDialogBody.OtherReferenceCount, crossDatasetReferences)))))), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: "none"
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, null, "If you ", action, " this document, documents that refer to it will no longer be able to access it.")));
}
