// @sanity/default-layout/lib/navbar/studioStatus/CurrentVersionsDialog.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _components = require("@sanity/base/components");

var _ui = require("@sanity/ui");

var _VersionsTable = require("./VersionsTable");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class CurrentVersionsDialog extends _react.default.PureComponent {
  renderTable() {
    var versions = this.props.versions;
    var rows = Object.keys(versions).map(pkgName => {
      return {
        name: pkgName,
        items: [versions[pkgName], versions[pkgName]]
      };
    });
    return /*#__PURE__*/_react.default.createElement(_VersionsTable.VersionsTable, {
      headings: ['Module', 'Installed', 'Latest'],
      rows: rows
    });
  }

  render() {
    var onClose = this.props.onClose;
    return /*#__PURE__*/_react.default.createElement(_components.LegacyLayerProvider, {
      zOffset: "navbarDialog"
    }, /*#__PURE__*/_react.default.createElement(_ui.Dialog, {
      header: "The Studio is up to date",
      id: "current-versions-dialog",
      onClose: onClose,
      onClickOutside: onClose,
      width: 1,
      scheme: "light"
    }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 4
    }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 5
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, null, "This Studio was built using the latest versions of all packages."), /*#__PURE__*/_react.default.createElement(_ui.Card, {
      as: "details",
      paddingTop: 4,
      borderTop: true
    }, /*#__PURE__*/_react.default.createElement("summary", null, "List all installed packages"), /*#__PURE__*/_react.default.createElement(_ui.Box, {
      marginTop: 4
    }, this.renderTable()))))));
  }

}

_defineProperty(CurrentVersionsDialog, "defaultProps", {
  versions: []
});

var _default = CurrentVersionsDialog;
exports.default = _default;
