// @sanity/default-layout/lib/navbar/studioStatus/SanityStatus.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = SanityStatus;

var _react = _interopRequireDefault(require("react"));

var _autoId = require("@reach/auto-id");

var _icons = require("@sanity/icons");

var _components = require("../components");

var _CurrentVersionsDialog = _interopRequireDefault(require("./CurrentVersionsDialog"));

var _UpdateNotifierDialog = _interopRequireDefault(require("./UpdateNotifierDialog"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function formatUpdateLabel(len) {
  if (len === 1) return ' 1 update';
  return "".concat(len, " updates");
}

function SanityStatus(props) {
  var isSupported = props.isSupported,
      isUpToDate = props.isUpToDate,
      level = props.level,
      onHideDialog = props.onHideDialog,
      onShowDialog = props.onShowDialog,
      _props$outdated = props.outdated,
      outdated = _props$outdated === void 0 ? [] : _props$outdated,
      showDialog = props.showDialog,
      versions = props.versions;
  var elementId = (0, _autoId.useId)();
  var currentLevel = outdated.length ? level : 'notice';
  var severity = isSupported ? currentLevel : 'high';
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, showDialog && /*#__PURE__*/_react.default.createElement("div", {
    role: "dialog",
    "aria-modal": "true",
    "aria-labelledby": elementId
  }, isUpToDate ? /*#__PURE__*/_react.default.createElement(_CurrentVersionsDialog.default, {
    onClose: onHideDialog,
    versions: versions
  }) : /*#__PURE__*/_react.default.createElement(_UpdateNotifierDialog.default, {
    onClose: onHideDialog,
    outdated: outdated,
    severity: severity
  })), !isUpToDate && /*#__PURE__*/_react.default.createElement(_components.StatusButton, {
    icon: _icons.PackageIcon,
    mode: "bleed",
    statusTone: "primary",
    selected: showDialog,
    "aria-label": "".concat(formatUpdateLabel(outdated.length), ", ").concat(severity, " severity level."),
    onClick: onShowDialog
  }));
}
