// @sanity/default-layout/lib/navbar/studioStatus/UpdateNotifierDialog.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _components = require("@sanity/base/components");

var _icons = require("@sanity/icons");

var _ui = require("@sanity/ui");

var _react = _interopRequireDefault(require("react"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _VersionsTable = require("./VersionsTable");

var _templateObject, _templateObject2;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var upperFirst = str => "".concat(str.slice(0, 1).toUpperCase()).concat(str.slice(1));

var ShellCode = (0, _styledComponents.default)(_ui.Code).attrs({
  language: 'shell'
})(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  & > code:before {\n    content: '$ ';\n  }\n"])));
var ToggleDetailsIcon = (0, _styledComponents.default)(_icons.ToggleArrowRightIcon)(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n  transition: transform 150ms;\n\n  details[open] & {\n    transform: rotate(90deg);\n  }\n"])));

class UpdateNotifierDialog extends _react.default.PureComponent {
  renderTable() {
    var outdated = this.props.outdated;
    var rows = outdated.map(pkg => {
      return {
        name: pkg.name,
        items: [pkg.version, pkg.latest, upperFirst(pkg.severity || 'low')]
      };
    });
    return /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 4
    }, /*#__PURE__*/_react.default.createElement(_VersionsTable.VersionsTable, {
      headings: ['Module', 'Installed', 'Latest', 'Importance'],
      rows: rows
    }), /*#__PURE__*/_react.default.createElement(_ui.Card, {
      padding: 4,
      radius: 2,
      tone: "primary"
    }, /*#__PURE__*/_react.default.createElement(_ui.Flex, null, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      size: 1
    }, /*#__PURE__*/_react.default.createElement(_icons.InfoOutlineIcon, null)), /*#__PURE__*/_react.default.createElement(_ui.Box, {
      flex: 1,
      marginLeft: 3
    }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      size: 1,
      weight: "semibold"
    }, "How to upgrade?"), /*#__PURE__*/_react.default.createElement(_ui.Text, {
      size: 1
    }, "Run the", ' ', /*#__PURE__*/_react.default.createElement("a", {
      href: "https://www.sanity.io/docs/reference/cli",
      target: "_blank",
      rel: "noreferrer"
    }, "Sanity CLI"), ' ', "upgrade command in the Terminal of your project directory:")), /*#__PURE__*/_react.default.createElement(_ui.Box, {
      marginTop: 4
    }, /*#__PURE__*/_react.default.createElement(ShellCode, null, "sanity upgrade"))))));
  }

  renderContactDeveloper() {
    var severity = this.props.severity;
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 4
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, null, severity === 'high' ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "This Studio should be updated. Please get in touch with the developers and ask them to upgrade it for you.") : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "This Studio has available upgrades. Consider getting in touch with the developers and ask them to upgrade it for you."))), /*#__PURE__*/_react.default.createElement(_ui.Card, {
      as: "details",
      borderTop: true,
      padding: 4
    }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
      as: "summary"
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, null, /*#__PURE__*/_react.default.createElement(ToggleDetailsIcon, null)), /*#__PURE__*/_react.default.createElement(_ui.Box, {
      flex: 1,
      marginLeft: 2
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      weight: "semibold"
    }, "Developer info"))), /*#__PURE__*/_react.default.createElement(_ui.Box, {
      marginTop: 4
    }, this.renderTable())));
  }

  render() {
    var _this$props = this.props,
        severity = _this$props.severity,
        onClose = _this$props.onClose;
    return /*#__PURE__*/_react.default.createElement(_components.LegacyLayerProvider, {
      zOffset: "navbarDialog"
    }, /*#__PURE__*/_react.default.createElement(_ui.Dialog, {
      header: severity === 'low' ? 'Upgrades available' : 'Studio is outdated',
      onClose: onClose,
      onClickOutside: onClose,
      id: "update-notifier-dialog",
      width: 1,
      scheme: "light"
    }, __DEV__ && /*#__PURE__*/_react.default.createElement(_ui.Box, {
      padding: 4
    }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 5
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, null, "This Studio is no longer up to date", ' ', severity === 'high' ? 'and should be upgraded.' : 'and can be upgraded.'), this.renderTable())), !__DEV__ && this.renderContactDeveloper()));
  }

}

_defineProperty(UpdateNotifierDialog, "defaultProps", {
  outdated: []
});

var _default = UpdateNotifierDialog;
exports.default = _default;
