// @sanity/default-layout/lib/navbar/studioStatus/VersionsTable.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VersionsTable = VersionsTable;

var _react = _interopRequireDefault(require("react"));

var _ui = require("@sanity/ui");

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _templateObject;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var CodeWithTextOverflowEllipsis = (0, _styledComponents.default)(_ui.Code)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  & > code {\n    display: block;\n    white-space: nowrap;\n    text-overflow: ellipsis;\n    overflow: hidden;\n  }\n"])));

function VersionsTable(props) {
  var headings = props.headings,
      rows = props.rows;

  if (rows.length === 0) {
    return null;
  }

  return /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 2,
    role: "table"
  }, /*#__PURE__*/_react.default.createElement(_ui.Flex, null, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1,
    weight: "semibold",
    role: "columnheader",
    textOverflow: "ellipsis"
  }, headings[0])), /*#__PURE__*/_react.default.createElement(_ui.Flex, {
    flex: 1
  }, headings.slice(1, headings.length).map(heading => /*#__PURE__*/_react.default.createElement(_ui.Box, {
    flex: 1,
    key: heading
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    size: 1,
    weight: "semibold",
    role: "columnheader",
    textOverflow: "ellipsis"
  }, heading))))), /*#__PURE__*/_react.default.createElement(_ui.Box, null, rows === null || rows === void 0 ? void 0 : rows.map(row => {
    var _row$items;

    return /*#__PURE__*/_react.default.createElement(_ui.Card, {
      key: row.name,
      display: "flex",
      paddingY: 2,
      role: "row"
    }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
      flex: 1
    }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
      flex: 1,
      role: "rowheader"
    }, /*#__PURE__*/_react.default.createElement(CodeWithTextOverflowEllipsis, null, row.name)), /*#__PURE__*/_react.default.createElement(_ui.Flex, {
      flex: 1
    }, row === null || row === void 0 ? void 0 : (_row$items = row.items) === null || _row$items === void 0 ? void 0 : _row$items.map(item => /*#__PURE__*/_react.default.createElement(_ui.Box, {
      flex: 1,
      role: "gridcell",
      key: item
    }, /*#__PURE__*/_react.default.createElement(CodeWithTextOverflowEllipsis, null, item))))));
  })));
}
