// @sanity/desk-tool/lib/components/MissingDocumentTypesMessage.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MissingDocumentTypesMessage = MissingDocumentTypesMessage;

var _react = _interopRequireDefault(require("react"));

var _ui = require("@sanity/ui");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function MissingDocumentTypesMessage() {
  return /*#__PURE__*/_react.default.createElement(_ui.Card, {
    height: "fill",
    paddingX: [5, 5, 7],
    paddingY: [5, 5, 6],
    sizing: "border"
  }, /*#__PURE__*/_react.default.createElement(_ui.Container, null, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
    space: 5
  }, /*#__PURE__*/_react.default.createElement(_ui.Heading, {
    as: "h1"
  }, "Empty schema"), /*#__PURE__*/_react.default.createElement(_ui.Text, {
    as: "p"
  }, "Your schema does not contain any document types. If it did, those types would be listed here.", ' ', /*#__PURE__*/_react.default.createElement("a", {
    title: "Schema documentation",
    target: "_blank",
    rel: "noopener noreferrer",
    href: "https://www.sanity.io/docs/content-studio/the-schema"
  }, "Read more about how to add schema types"), "."))));
}
