// @sanity/base/lib/datastores/document/document-pair/editState.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.editState = void 0;

var _operators = require("rxjs/operators");

var _createMemoizer = require("../utils/createMemoizer");

var _isLiveEditEnabled = require("./utils/isLiveEditEnabled");

var _operationArgs = require("./operationArgs");

var editState = (0, _createMemoizer.memoize)((idPair, typeName) => {
  var liveEdit = (0, _isLiveEditEnabled.isLiveEditEnabled)(typeName);
  return (0, _operationArgs.operationArgs)(idPair, typeName).pipe((0, _operators.map)(_ref => {
    var snapshots = _ref.snapshots;
    return {
      id: idPair.publishedId,
      type: typeName,
      draft: snapshots.draft,
      published: snapshots.published,
      liveEdit,
      ready: true
    };
  }), (0, _operators.startWith)({
    id: idPair.publishedId,
    type: typeName,
    draft: null,
    published: null,
    liveEdit,
    ready: false
  }), (0, _operators.publishReplay)(1), (0, _operators.refCount)());
}, (idPair, typeName) => idPair.publishedId + typeName);
exports.editState = editState;
